packages/@aws-cdk/aws-ec2/lib/cfn-init-elements.ts

```diff
@@ -0,0 +1,405 @@
+   */
+  readonly instanceRole: iam.IRole;
+}
+
+/**
+ * A return type for a configured InitElement. Both its CloudFormation representation, and any
+ * additional metadata needed to create the CloudFormation:Init.
+ */
+export interface InitElementConfig {
+  /**
+   * The CloudFormation representation of the configuration of an InitElement.
+   */
+  readonly config: Record<string, any>;
+
+  /**
+   * Optional authentication blocks to be associated with the Init Config
+   *
+   * @default - No authentication associated with the config
+   */
+  readonly authentication?: Record<string, any>;
+}
+
+/**
+ * Base class for all CloudFormation Init elements
+ */
+export abstract class InitElement {
+
+  /**
+   * Returns the init element type for this element.
+   */
+  public abstract readonly elementType: InitElementType;
+
+  /**
+   * Called when the Init config is being consumed. Renders the CloudFormation
+   * representation of this init element, and calculates any authentication
+   * properties needed, if any.
+   *
+   * @param options bind options for the element.
+   */
+  public abstract bind(options: InitBindOptions): InitElementConfig;
+
+}
+
+/**
+ * Options for InitCommand
+ */
+export interface InitCommandOptions {
+  /**
+   * Identifier key for this command
+   *
+   * You can use this to order commands.
+   *
+   * @default - Automatically generated
+   */
+  readonly key?: string;
+
+  /**
+   * Sets environment variables for the command.
+   *
+   * This property overwrites, rather than appends, the existing environment.
+   *
+   * @default - Use current environment
+   */
+  readonly env?: Record<string, string>;
+
+  /**
+   * The working directory
+   *
+   * @default - Use default working directory
+   */
+  readonly cwd?: string;
+
+  /**
+   * Command to determine whether this command should be run
+   *
+   * If the test passes (exits with error code of 0), the command is run.
+   *
+   * @default - Always run this command
+   */
+  readonly test?: string;
+
+  /**
+   * Continue running if this command fails
+   *
+   * @default false
+   */
+  readonly ignoreErrors?: boolean;
+
+  /**
+   * Specifies how long to wait (in seconds) after a command has finished in case the command causes a reboot.
+   *
+   * A value of "forever" directs cfn-init to exit and resume only after the
+   * reboot is complete. Set this value to 0 if you do not want to wait for
+   * every command.
+   *
+   * For Windows systems only.
+   *
+   * @default Duration.seconds(60)
+   */
+  readonly waitAfterCompletion?: Duration;
+
+  /**
+   * Restart the given service(s) after this command has run
+   *
+   * @default - Do not restart any service
+   */
+  readonly serviceRestartHandles?: InitServiceRestartHandle[];
+}
+
+/**
+ * Command to execute on the instance
+ */
+export class InitCommand extends InitElement {
+  /**
+   * Run a shell command
+   *
+   * You must escape the string appropriately.
```

Comment:
What needs to be escaped? Can we do this automatically for the user?

Reply:
> Maybe better to just remove this comment entirely?

If you've tested this at some basic level, I can buy that. 

We can add a different method here when we realize there is escaping we've missed and deprecate the existing API.

Reply:
This comment is not addressed. 

As discussed, remove the statement around escaping.

Reply:
The comment here is about shell escaping. It's if you take user input, people simply append strings and get a red face when someone inputs `"; rm -rf /"` as a input.

This warning should be on every string that goes into a shell to remind people that sharp edges are sharp.